### ISSUE TITLE
cannot import name 'rcParams' from 'matplotlib.axes'

### ISSUE THREAD

**Issue by programmeddeath1** [pyinstaller/pyinstaller#8697]
This is my .spec file 
```
# -*- mode: python ; coding: utf-8 -*-
import os
import sys
from PyInstaller.utils.hooks import collect_data_files, collect_dynamic_libs, copy_metadata

# Add ROS paths
ros_paths = [

    '/opt/ros/noetic/lib/python3/dist-packages',
]

# Extend the Python path
sys.path.extend(ros_paths)

# Collect data files for pygraphviz
datas = collect_data_files('pygraphviz')
datas += collect_data_files('boxmot')
datas += collect_data_files('matplotlib', subdir='mpl-data')

# Additional binaries for Graphviz
graphviz_binaries = [
    ('/usr/bin/neato', 'graphviz/bin'),
    ('/usr/bin/dot', 'graphviz/bin'),
    ('/usr/bin/fdp', 'graphviz/bin'),
    ('/usr/bin/sfdp', 'graphviz/bin'),
    ('/usr/bin/twopi', 'graphviz/bin'),
    ('/usr/bin/circo', 'graphviz/bin'),
]

#binaries.extend(graphviz_binaries)

# Collect metadata for python imports
datas += copy_metadata('norfair')
datas += copy_metadata('boxmot')
datas += copy_metadata('depthai')
datas += copy_metadata('matplotlib')
datas += copy_metadata('ultralytics')

# Include additional data files
datas += [
    ('config/delta_robot.rviz', 'config'),
    ('images/camera.png', 'images'),
    ('images/connections.png', 'images'),
    ('images/control-centre.png', 'images'),
    ('images/green-light.png', 'images'),
    ('images/home.png', 'images'),
    ('images/red-light.png', 'images'),
    ('images/screen3.png', 'images'),
    ('images/settings.png', 'images'),
    ('images/load_spinner.gif','images'),
    ('libraries/calc.py', 'libraries'),
    ('libraries/__init__.py', 'libraries'),
    ('model/bus.jpg', 'model'),
    ('model/coco_80_labels_list.txt', 'model'),
    ('static/one.ui', 'static'),
    ('static/two.ui', 'static'),
    ('static/three.ui', 'static'),
    ('static/four.ui', 'static'),
    ('executor.py', '.'),
    ('rknn_yolov5_demo', '.'),
    ('rknn_yolov5_demo_dynamichome', '.'),
    ('runmodel_server.sh', '.'),
    ('image.jpg', '.'),
    ('testing/kitkat_trid_video_stream.mp4', 'testing'),
    ('config/workspace.json', 'config'),
    ('config/botsort.yaml', 'config'),
    ('config/camera_detection.yaml', 'config'),
    ('config/config.myviz', 'config'),
    ('config/config.yaml', 'config'),
    ('config/log_config.yaml', 'config'),
    ('config/speed_limits.csv', 'config'),
]


a = Analysis(
    ['main_cpu.py'],
    pathex=ros_paths + sys.path,
    binaries=graphviz_binaries,
    datas=datas,
    hiddenimports=[
        'roslib',
        'rospy',
        'std_msgs',
        'geometry_msgs',
        'sensor_msgs',
        'norfair',
        'boxmot',
        'ultralytics',
        'custom_msg.srv.Trajectory',
        'custom_msg.srv.TrajectoryResponse',
        'custom_msg.srv.TrajectoryRequest',
        'custom_msg.msg.ObjectPose',
        'custom_msg.msg.CameraStreamMsg',
        'custom_msg.msg.JoggingMsg',
        'matplotlib',  # Ensure matplotlib is included
        'mpl_toolkits.mplot3d'  # Ensure mpl_toolkits.mplot3d is included
    ],
    hookspath=['.'],
    hooksconfig={},
    runtime_hooks=[],
    #excludes=['PySide2'],
    excludes=['PySide2', 'lz4'],
    noarchive=False,
    optimize=0,
)

pyz = PYZ(a.pure)

exe = EXE(
    pyz,
    a.scripts,
    [],
    exclude_binaries=True,
    name='mainapp',
    debug=False,
    bootloader_ignore_signals=False,
    strip=False,
    upx=True,
    console=True,
    disable_windowed_traceback=False,
    argv_emulation=False,
    target_arch=None,
    codesign_identity=None,
    entitlements_file=None,
)

coll = COLLECT(
    exe,
    a.binaries,
    a.datas,
    strip=False,
    upx=True,
    upx_exclude=[],
    name='mainapp',
)
```

the app builds fine but wen it runs from ultralytics import YOLO it fails with 
```
Default Model set is - yolov8s.pt
Traceback (most recent call last):
  File "main_cpu.py", line 6002, in <module>
    w.addWidget(four(ros_listener_thread=ros_listener_thread))
  File "main_cpu.py", line 5079, in __init__
    self.load_settings()
  File "main_cpu.py", line 5104, in load_settings
    self.populate_models_combobox()
  File "main_cpu.py", line 5406, in populate_models_combobox
    self.select_default_model(default_model_path) 
  File "main_cpu.py", line 5176, in select_default_model
    self.set_global_model_data(default_model,call_instance="init")            
  File "main_cpu.py", line 5426, in set_global_model_data
    from ultralytics import YOLO
  File "PyInstaller/loader/pyimod02_importers.py", line 419, in exec_module
  File "ultralytics/__init__.py", line 5, in <module>
  File "PyInstaller/loader/pyimod02_importers.py", line 419, in exec_module
  File "ultralytics/models/__init__.py", line 3, in <module>
  File "PyInstaller/loader/pyimod02_importers.py", line 419, in exec_module
  File "ultralytics/models/rtdetr/__init__.py", line 3, in <module>
  File "PyInstaller/loader/pyimod02_importers.py", line 419, in exec_module
  File "ultralytics/models/rtdetr/model.py", line 10, in <module>
  File "PyInstaller/loader/pyimod02_importers.py", line 419, in exec_module
  File "ultralytics/engine/model.py", line 8, in <module>
  File "PyInstaller/loader/pyimod02_importers.py", line 419, in exec_module
  File "ultralytics/cfg/__init__.py", line 10, in <module>
  File "PyInstaller/loader/pyimod02_importers.py", line 419, in exec_module
  File "ultralytics/utils/__init__.py", line 19, in <module>
  File "PyInstaller/loader/pyimod02_importers.py", line 419, in exec_module
  File "matplotlib/pyplot.py", line 61, in <module>
  File "PyInstaller/loader/pyimod02_importers.py", line 419, in exec_module
  File "matplotlib/figure.py", line 43, in <module>
  File "PyInstaller/loader/pyimod02_importers.py", line 419, in exec_module
  File "matplotlib/projections/__init__.py", line 58, in <module>
  File "PyInstaller/loader/pyimod02_importers.py", line 419, in exec_module
  File "mpl_toolkits/mplot3d/__init__.py", line 1, in <module>
  File "PyInstaller/loader/pyimod02_importers.py", line 419, in exec_module
  File "mpl_toolkits/mplot3d/axes3d.py", line 27, in <module>
ImportError: cannot import name 'rcParams' from 'matplotlib.axes' (/home/-/app/delta-gui/dist/mainapp/_internal/matplotlib/axes/__init__.pyc)
[118984] Failed to execute script 'main_cpu' due to unhandled exception!
```
I have added matplotlib data and even added a hook as below - 
```from PyInstaller.utils.hooks import collect_data_files, collect_submodules

# Collect data files from matplotlib
datas = collect_data_files('matplotlib', subdir='mpl-data')

# Collect all submodules from matplotlib
hiddenimports = collect_submodules('matplotlib')

# Collect all submodules from mpl_toolkits.mplot3d
hiddenimports += collect_submodules('mpl_toolkits.mplot3d')

# Collect data files from mpl_toolkits.mplot3d
datas += collect_data_files('mpl_toolkits.mplot3d')
```

But it keeps giving this error, and the matplot lib folder in _internal does not contain an axes folder it only creates the following 
backends
_c_internal_utils.cpython-38-x86_64-linux-gnu.so
ft2font.cpython-38-x86_64-linux-gnu.so
_image.cpython-38-x86_64-linux-gnu.so
mpl-data
_path.cpython-38-x86_64-linux-gnu.so
_qhull.cpython-38-x86_64-linux-gnu.so
_tri.cpython-38-x86_64-linux-gnu.so
_ttconv.cpython-38-x86_64-linux-gnu.so

**Comment by rokm** [comment]
What version of `matplotlib` are you using?

**Comment by rokm** [comment]
> and the matplot lib folder in _internal does not contain an axes folder

Why would it? If `matplotlib.axes` contains only pure-python modules, it is collected only into PYZ archive.

If you want to properly debug what is included and what not, you will either need to take a look at `build/<name>/Analysis-00.toc` or `build/<name>/PYZ-00.toc`, or open the executable with `pyi-archive_viewer` to open the embedded PKG archive, and then open the PYZ archive entry from it.

**Comment by rokm** [comment]
Either way, my first guess here would be that you ended up with a different version of `mpl_toolkits` and of `matplotlib` collected (from two different locations), and the two of them are incompatible. You can verify that by checking the afore-mentioned `build/<name>/Analysis-00.toc` - specifically, check where `matplotlib` modules are being collected from, and where `mpl_toolkits` ones are being collected from.

**Comment by programmeddeath1** [comment]
I have tried with 3.7.5 and 3.6 and 3.5 versions. And this is actually my original spec file - 
```
# -*- mode: python ; coding: utf-8 -*-
import os
import sys
from PyInstaller.utils.hooks import collect_data_files, collect_dynamic_libs, copy_metadata

# Add ROS paths
ros_paths = [
    '/home/greenpi/noetic_ws/install_isolated/lib/python3/dist-packages',
    '/home/greenpi/noetic_ws/devel_isolated/lib/python3/dist-packages'
]

# Extend the Python path
sys.path.extend(ros_paths)

# Collect data files for pygraphviz
datas = collect_data_files('pygraphviz')
datas += collect_data_files('boxmot')
#datas = collect_data_files('matplotlib')

# Collect dynamic libraries for lz4
#binaries = collect_dynamic_libs('lz4')

# Additional binaries for Graphviz
graphviz_binaries = [
    ('/usr/bin/neato', 'graphviz/bin'),
    ('/usr/bin/dot', 'graphviz/bin'),
    ('/usr/bin/fdp', 'graphviz/bin'),
    ('/usr/bin/sfdp', 'graphviz/bin'),
    ('/usr/bin/twopi', 'graphviz/bin'),
    ('/usr/bin/circo', 'graphviz/bin'),
]

#binaries.extend(graphviz_binaries)

# Collect metadata for python imports
datas += copy_metadata('norfair')
datas += copy_metadata('boxmot')
datas += copy_metadata('depthai')
datas += copy_metadata('ultralytics')

# Include additional data files
datas += [
    ('config/delta_robot.rviz', 'config'),
    ('images/camera.png', 'images'),
    ('images/connections.png', 'images'),
    ('images/control-centre.png', 'images'),
    ('images/green-light.png', 'images'),
    ('images/home.png', 'images'),
    ('images/red-light.png', 'images'),
    ('images/screen3.png', 'images'),
    ('images/settings.png', 'images'),
    ('images/load_spinner.gif','images'),
    ('lib/libmk_api.so', 'lib'),
    ('libraries/__init__.py', 'libraries'),
    ('model/bus.jpg', 'model'),
    ('model/coco_80_labels_list.txt', 'model'),
    ('static/one.ui', 'static'),
    ('static/two.ui', 'static'),
    ('static/three.ui', 'static'),
    ('static/four.ui', 'static'),
    ('executor.py', '.'),
    ('runmodel_server.sh', '.'),
    ('image.jpg', '.'),
    ('testing/kitkat_trid_video_stream.mp4', 'testing'),
    ('config/workspace.json', 'config'),
    ('config/botsort.yaml', 'config'),
    ('config/camera_detection.yaml', 'config'),
    ('config/config.myviz', 'config'),
    ('config/config.yaml', 'config'),
    ('config/log_config.yaml', 'config'),
    ('config/speed_limits.csv', 'config'),
]

# Hook file for matplotlib
#hook_file_content = """
#from PyInstaller.utils.hooks import collect_data_files, collect_submodules
#
#datas = collect_data_files('matplotlib')
#hiddenimports = collect_submodules('matplotlib')
#"""
#
## Write the hook file
#with open('hook-matplotlib.py', 'w') as hook_file:
#    hook_file.write(hook_file_content)

a = Analysis(
    ['main.py'],
    pathex=ros_paths + sys.path,
    binaries=graphviz_binaries,
    datas=datas,
    hiddenimports=[
        'roslib',
        'rospy',
        'std_msgs',
        'geometry_msgs',
        'sensor_msgs',
        'norfair',
        'boxmot',
        'ultralytics',
        'custom_msg.srv.Trajectory',
        'custom_msg.srv.TrajectoryResponse',
        'custom_msg.srv.TrajectoryRequest',
        'custom_msg.msg.ObjectPose',
        'custom_msg.msg.CameraStreamMsg',
        'custom_msg.msg.JoggingMsg'        
        # Add other ROS packages that your application depends on
    ],
    hookspath=['.'],
    hookspath=['.'],
    hooksconfig={},
    runtime_hooks=[],
    #excludes=['PySide2'],
    excludes=['PySide2', 'lz4'],
    noarchive=False,
    optimize=0,
)

pyz = PYZ(a.pure)

exe = EXE(
    pyz,
    a.scripts,
    [],
    exclude_binaries=True,
    name='mainapp',
    debug=False,
    bootloader_ignore_signals=False,
    strip=False,
    upx=True,
    console=True,
    disable_windowed_traceback=False,
    argv_emulation=False,
    target_arch=None,
    codesign_identity=None,
    entitlements_file=None,
)

coll = COLLECT(
    exe,
    a.binaries,
    a.datas,
    strip=False,
    upx=True,
    upx_exclude=[],
    name='mainapp',
)

```
I had added the matplotlib as hidden import and hook/collect_data_files directly only after this default spec file started giving me  the rcparams error. That error persists even after adding matplotlib data using collect_data.

So I also checked the Analysis-00.toc. Matplotlib imports are all from the same path - .local/lib/python3.8/site-packages/matplotlib

mpl_toolkits imports are from the path - /usr/lib/python3/dist-packages/mpl_toolkits/

[here](https://github.com/programmeddeath1/webhost/blob/master/matployt-Analysis-00.toc) is a shortened version of the Analysis-00.toc with the matplotlib imports

**Comment by rokm** [comment]
> So I also checked the Analysis-00.toc. Matplotlib imports are all from the same path - .local/lib/python3.8/site-packages/matplotlib
>
> mpl_toolkits imports are from the path - /usr/lib/python3/dist-packages/mpl_toolkits/

That looks suspicious, and is likely the cause of the issue - both `matplotlib` and `mpl_toolkits` importable package should come from the same `matplotlib` dist, and should therefore come from the same location. 

And just to confirm this, can you check that there also exists a `.local/lib/python3.8/site-packages/mpl_toolkits` directory, and similarly a `/usr/lib/python3/dist-packages/matplotlib` directory ?

**Comment by programmeddeath1** [comment]
yeah it does exist. there is a matplotlib and mpl_toolkits/ installation in user python directory as well as in installation in user python directory as well as /usr/lib/python3. 
so both ls /usr/lib/python3/dist-packages/mpl_toolkits/ and /usr/lib/python3/dist-packages/matplotlib/ exist along with  ~/.local/lib/python3.8/site-packages/matplotlib and ~/.local/lib/python3.8/site-packages/mpl_toolkits/
```
>>> import matplotlib as mp
>>> mp.__path__
['/home/_user_/.local/lib/python3.8/site-packages/matplotlib']
>>> import mpl_toolkits as mpl
>>> mpl.__path__
_NamespacePath(['/home/_user_/.local/lib/python3.8/site-packages/mpl_toolkits', '/usr/lib/python3/dist-packages/mpl_toolkits'])
```

**Comment by rokm** [comment]
Then you have a system-installed matplotlib in `/usr/lib/python3/dist-packages` and pip-installed `matplotlib` in user site-pacakges (`/home/_user_/.local/lib/python3.8/site-packages`), and the two are incompatible.

You should sort out your search paths to avoid having multiple versions of the same package, as this can confuse PyInstaller's import analysis. And in general, you should avoid using user site-packages location; if you need to install packages via pip, create and use a virtual environment instead. In general, you are better off with using minimal virtual environments when using PyInstaller.

**Comment by programmeddeath1** [comment]
i uninstalled the user installations and continue with the system installations.  it still fails with the same error when running the executable.
```
Default Model set is - yolov8s.pt
Traceback (most recent call last):
  File "main_cpu.py", line 6002, in <module>
    w.addWidget(four(ros_listener_thread=ros_listener_thread))
  File "main_cpu.py", line 5079, in __init__
    self.load_settings()
  File "main_cpu.py", line 5104, in load_settings
    self.populate_models_combobox()
  File "main_cpu.py", line 5406, in populate_models_combobox
    self.select_default_model(default_model_path) 
  File "main_cpu.py", line 5176, in select_default_model
    self.set_global_model_data(default_model,call_instance="init")            
  File "main_cpu.py", line 5426, in set_global_model_data
    from ultralytics import YOLO
  File "PyInstaller/loader/pyimod02_importers.py", line 419, in exec_module
  File "ultralytics/__init__.py", line 5, in <module>
  File "PyInstaller/loader/pyimod02_importers.py", line 419, in exec_module
  File "ultralytics/models/__init__.py", line 3, in <module>
  File "PyInstaller/loader/pyimod02_importers.py", line 419, in exec_module
  File "ultralytics/models/rtdetr/__init__.py", line 3, in <module>
  File "PyInstaller/loader/pyimod02_importers.py", line 419, in exec_module
  File "ultralytics/models/rtdetr/model.py", line 10, in <module>
  File "PyInstaller/loader/pyimod02_importers.py", line 419, in exec_module
  File "ultralytics/engine/model.py", line 8, in <module>
  File "PyInstaller/loader/pyimod02_importers.py", line 419, in exec_module
  File "ultralytics/cfg/__init__.py", line 10, in <module>
  File "PyInstaller/loader/pyimod02_importers.py", line 419, in exec_module
  File "ultralytics/utils/__init__.py", line 19, in <module>
  File "PyInstaller/loader/pyimod02_importers.py", line 419, in exec_module
  File "matplotlib/pyplot.py", line 61, in <module>
  File "PyInstaller/loader/pyimod02_importers.py", line 419, in exec_module
  File "matplotlib/figure.py", line 43, in <module>
  File "PyInstaller/loader/pyimod02_importers.py", line 419, in exec_module
  File "matplotlib/projections/__init__.py", line 58, in <module>
  File "PyInstaller/loader/pyimod02_importers.py", line 419, in exec_module
  File "mpl_toolkits/mplot3d/__init__.py", line 1, in <module>
  File "PyInstaller/loader/pyimod02_importers.py", line 419, in exec_module
  File "mpl_toolkits/mplot3d/axes3d.py", line 27, in <module>
ImportError: cannot import name 'rcParams' from 'matplotlib.axes' (/home/_user_/app/delta-gui/dist/mainapp/_internal/matplotlib/axes/__init__.pyc)
[23104] Failed to execute script 'main_cpu' due to unhandled exception!
```

[here](https://github.com/programmeddeath1/webhost/blob/master/matployt-Analysis-00.toc) is the Analysis file
I think the matplotlib folder in _internal is not getting created properly , which could probably point to data not being moved or imported properly

**Comment by rokm** [comment]
Hmm, okay. What distribution is that, and what is the version of system-installed `matplotlib` that you are using now?

How about the following test program:

```python
from matplotlib.axes import rcParams
print("hello")
```

Does this script work if you run it under python? Does it work if you freeze it using PyInstaller (without any additional options or spec customization)?

**Comment by rokm** [comment]
And just to be sure it is not a cache-related issue now, have you tried rebuilding with added `--clean` option?

**Comment by programmeddeath1** [comment]
Yeah matplotlib.axes import rcParams works under both python and even as Pyinstaller. i am running with --clean.
The problem is at the line - from ultralytics import YOLO in my app.
Ultralytics requires matplotlib from the user installation. pip install ultralytics runs pip install matplotlib.

Thus when I uninstalled user matplotlib and ran the app, it fails with - FileNotFoundError: [Errno 2] No such file or directory: '/home/_user_/app/delta-gui/dist/mainapp/_internal/ultralytics/cfg/default.yaml' because now since ultralytics does not get matplotlib, it is not properly imported by pyistaller.

The system matplotlib installation is done by the ROS package which uses the system python packages for its dependencies. 

I will have to try to see if I  bring the ROS dependencies to the ROS python dist packages and not read it from system packages.

Is there a way to make a specific package be imported from user location and skip its system path import  or vice versa in its hook?

**Comment by rokm** [comment]
> Is there a way to make a specific package be imported from user location and skip its system path import or vice versa in its hook?

No. But do you actually need anything from system python installation? If not, use a clean virtual environment, and pip-install your dependencies into that.

**Comment by programmeddeath1** [comment]
yeah the pyqt5 and ros packages use system packages. I had actually started the prject with venv because thats the default I work with, but ROS 1 uses pyqt python bindings which have a dependency with the system installed pyqt5. I had tried with venv installed pyqt5 but was facing a myriad of issues trying to fix this dependency so had resorted to using system installations of all python packages.

It should be possible but that looks like an arduous task, but since i seemed to have hit a brick wall here, I might have to try and get to the bottom of that.

Thank you for your help! much appreciated
